File: solutions/cpp/0018.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> ans;
        vector<int> path;
        sort(nums.begin(), nums.end());
        nSum(nums, target, 4, path, ans);
        return ans;
    }

private:
    void nSum(vector<int>& nums, int target, int n, vector<int>& path, vector<vector<int>>& ans) {
        if (nums.size() < n || n < 2 || target < nums[0] * n || target > nums.back() * n)
            return;

        if (n == 2) {
            int l = 0;
            int r = nums.size() - 1;
            while (l < r) {
                if (nums[l] + nums[r] == target) {
                    vector<int> vec(path);
                    vec.insert(vec.end(), {nums[l], nums[r]});
                    ans.push_back(vec);
                    l++;
                    while (l < r && nums[l] == nums[l - 1]) l++;
                } else if (nums[l] + nums[r] < target) {
                    l++;
                } else {
                    r--;
                }
            }
        } else {
            for (int i = 0; i < nums.size() - n + 1; i++) {
                if (i == 0 || i > 0 && nums[i - 1] != nums[i]) {
                    vector<int>::const_iterator first = nums.begin() + i + 1;
                    vector<int>::const_iterator last = nums.end();
                    vector<int> newNums(first, last);
                    path.push_back(nums[i]);
                    nSum(newNums, target - nums[i], n - 1, path, ans);
                    path.pop_back();
                }
            }
        }
    }
};

File: solutions/python3/0018.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans = []
        self.nSum(sorted(nums), target, 4, [], ans)
        return ans

    def nSum(self, nums: List[int], target: int, n: int, curr: List[int], ans: List[List[int]]) -> None:
        if len(nums) < n or n < 2 or target < nums[0] * n or target > nums[-1] * n:
            return

        if n == 2:
            l = 0
            r = len(nums) - 1
            while l < r:
                if nums[l] + nums[r] == target:
                    ans.append(curr + [nums[l], nums[r]])
                    l += 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
                elif nums[l] + nums[r] < target:
                    l += 1
                else:
                    r -= 1
        else:
            for i in range(len(nums) - n + 1):
                if i == 0 or i > 0 and nums[i - 1] != nums[i]:
                    self.nSum(nums[i + 1:], target - nums[i],
                              n - 1, curr + [nums[i]], ans)
